### PR TITLE
bare_root_metadata: handle upconversion to v3

### PR DESCRIPTION
**Pull request by aalness**
This handles converting from MDv2 to MDv3.

### CODE REVIEW

**Comment by aalness** [comment]
Just realized the key cache is causing my tests to not actually hit all of the code I want to test. Will fix.

File: libkbfs/key_bundle.go [hunk 1]
```diff
@@ -150,6 +170,57 @@ func (tkg TLFWriterKeyGenerations) IsWriter(user keybase1.UID, deviceKID keybase
...
+		// Sanity check.
+		if len(keys) != int(keyGen) {
+			return nil, fmt.Errorf("Expected %d keys, found %d\n", keyGen, len(keys))
+		}
+		// Save the current key.
+		currKey := keys[len(keys)-1]
+		// Get rid of the most current generation as that's in the UserDeviceKeyInfoMap already.
+		keys = keys[:len(keys)-1]
+		// Encrypt the historic keys with the current key.
+		wkbCopy.EncryptedHistoricTLFCryptKeys, err = crypto.EncryptTLFCryptKeys(keys, currKey)
```
Comment (jzila) [inline libkbfs/key_bundle.go]: :joy:

File: libkbfs/key_bundle.go [hunk 2]
```diff
@@ -193,6 +264,72 @@ func (tkg TLFReaderKeyGenerations) IsReader(user keybase1.UID, deviceKID keybase
...
+				}
+				// Map the old index in the writer list to a new index
+				// at the end of the reader list.
+				if newIndex, ok := pubKeyIndicesMap[oldIndex]; !ok {
+					ePubKey := wkb.TLFEphemeralPublicKeys[oldIndex]
+					rkbCopy.TLFReaderEphemeralPublicKeys =
+						append(rkbCopy.TLFReaderEphemeralPublicKeys, ePubKey)
+					newIndex = len(rkbCopy.TLFReaderEphemeralPublicKeys) - 1
+					pubKeyIndicesMap[oldIndex] = newIndex
+					info.EPubKeyIndex = newIndex
```
Comment (jzila) [inline libkbfs/key_bundle.go]: square peg round hole

File: libkbfs/root_metadata_test.go [hunk 3]
```diff
@@ -535,7 +535,186 @@ func TestRootMetadataFinalIsFinal(t *testing.T) {
...
+	config.KeybaseService().(*KeybaseDaemonLocal).addNewAssertionForTestOrBust(
+		"charlie", "charlie@twitter")
+
+	// rekey it
+	done, _, err = config.KeyManager().Rekey(context.Background(), rmd, false)
+	require.NoError(t, err)
+	require.True(t, done)
+	require.Equal(t, rmd.LatestKeyGeneration(), KeyGen(1))
+	require.Equal(t, rmd.Revision(), MetadataRevision(1))
+	require.Equal(t, rmd.Version(), InitialExtraMetadataVer)
```
Comment (jzila) [inline libkbfs/root_metadata_test.go]: Should check to make sure that there's a new TLFEphemeralPublicKey in the reader key list.

File: libkbfs/bare_root_metadata_v2.go [hunk 4]
```diff
@@ -52,6 +52,45 @@ type WriterMetadataV2 struct {
...
+		wmdCopy.LatestKeyGen = PublicKeyGen
+	} else {
+		wmdCopy.LatestKeyGen = wmd.WKeys.LatestKeyGeneration()
+	}
+	wmdCopy.ID = wmd.ID
+	wmdCopy.BID = wmd.BID
+	wmdCopy.WFlags = wmd.WFlags
+	wmdCopy.DiskUsage = wmd.DiskUsage
+	wmdCopy.RefBytes = wmd.RefBytes
+	wmdCopy.UnrefBytes = wmd.UnrefBytes
```
Comment (aalness) [inline libkbfs/bare_root_metadata_v2.go]: They're numbers not actual byte arrays.

**Pull request merged**